### DISCUSSION TITLE
Storing Object & Array in Preference

### DISCUSSION THREAD

**Discussion by maheshrajrp** [eclipse-theia/theia#10130] [General] [answered]
Hi,
I'm currently trying to store object with keys and values also arrays within a preference. Is it possible ? 
Also, I found editor.suggest.filteredTypes in preference, which can be similar to my usecase. However, I'm unable to find its implimentation. It would be valuable if you could provide a summary of what it does and how I could use it. 

The GUI part is not needed. The usecase would be fine with "Edit in settings.json"

Thanks !!

**Comment by vince-fugnitto** [comment, marked as answer]
@maheshrajrp you can take a look at `editor.quickSuggestions` as an example:

https://github.com/eclipse-theia/theia/blob/86181832c33b10e54117490b3f5a5bf9d13e142a/packages/editor/src/browser/editor-preferences.ts#L913-L945

In the `settings.json` it will look like:

```json
{
    "editor.quickSuggestions": {
        "other": true,
        "comments": false,
        "strings": false
    }
}
```

**Reply by maheshrajrp** [reply]
Hi @vince-fugnitto ,
Thanks !! I'll try this. Also, My Custom Preferences are being added within Extensions in Preferences. Is there any way to bring them as a parent category like it used be earlier (v.1.13) ?

**Reply by vince-fugnitto** [reply]
@maheshrajrp you'll have to modify how these categories are implemented:

https://github.com/eclipse-theia/theia/blob/86181832c33b10e54117490b3f5a5bf9d13e142a/packages/preferences/src/browser/util/preference-tree-generator.ts#L41-L50

**Reply by maheshrajrp** [reply]
Hi @vince-fugnitto ,
Just a few more doubts. 
```
 'editor.quickSuggestions':  { 
     'anyOf': [ 
         { 
             'type': 'boolean' 
         ...
```
1. What does this anyOf and the first 'type' mean ? Since, there isn't any key other than 'type' present in that I'm unable to understand why do we need an object having 'type' as the first object in this array.

2. Then, can this implementation be used to store array ? If yes  can you provide an example ?
 
```
 protected readonly topLevelCategories = new Map([ 
     [COMMONLY_USED_SECTION_PREFIX, 'Commonly Used'], 
     ['editor', 'Text Editor'], 
     ['workbench', 'Workbench'], 
     ['window', 'Window'], 
     ['features', 'Features'], 
     ['application', 'Application'], 
     ['extensions', 'Extensions'] 
```
3. I'm unable to understand how this can be used to add my custom preference as a parent. Like any examples or document would be helpful.

And definitely thanks for your time. I'm like posting questions almost everyday and you guys are kind enough to provide the solution. I really wish I could contribute to this project and am waiting for a good first issue that I could help out with (I'm really having hard time understanding even the good first issues as I've just began working/learning js, ts, react, etc).

**Reply by vince-fugnitto** [reply]
@maheshrajrp 

> 1. What does this anyOf and the first 'type' mean ? 

It's just an example, but if you want more information you can look at the _json schema_ documentation:

- https://json-schema.org/understanding-json-schema/reference/combining.html#anyof

> 2. Then, can this implementation be used to store array ? If yes can you provide an example ?

Which implementation exactly? `arrays` are valid types and there are multiple examples if you search the code-base:

https://github.com/eclipse-theia/theia/blob/86181832c33b10e54117490b3f5a5bf9d13e142a/packages/editor/src/browser/editor-preferences.ts#L1035-L1042

> 3. I'm unable to understand how this can be used to add my custom preference as a parent. Like any examples or document would be helpful.

The snippet is the example. You'll have to see how it is being used, and modify it to suit your needs. If you want your preference as a top-level element in the tree you'll have to implement it that way.

**Reply by maheshrajrp** [reply]
Thanks Again @vince-fugnitto !! I'm able to understand now.
